src/form/wr.jl

```diff
@@ -736,3 +737,97 @@ function constraint_power_magnitude_link_on_off(pm::QCWRPowerModel, n::Int, i, a
+function QCWRTriPowerModel(data::Dict{String,Any}; kwargs...)
+    return GenericPowerModel(data, QCWRTriForm; kwargs...)
+end
+
+""
+function variable_voltage_magnitude_product{T <: QCWRTriForm}(pm::GenericPowerModel{T}, n::Int=pm.cnw)
+    # do nothing - no lifted variables required for voltage variable product
+end 
+
+"creates lambda variables for convex combination model" 
+function variable_multipliers{T <: QCWRTriForm}(pm::GenericPowerModel{T}, n::Int=pm.cnw)
+    buspairs = pm.ref[:nw][n][:buspairs]
+    pm.var[:nw][n][:lambdar] = @variable(pm.model, 
+        [bp in keys(pm.ref[:nw][n][:buspairs]), i=1:8], basename="$(n)_lambda",
+        lowerbound = 0, upperbound = 1)
+    pm.var[:nw][n][:lambdai] = @variable(pm.model, 
+        [bp in keys(pm.ref[:nw][n][:buspairs]), i=1:8], basename="$(n)_lambda",
+        lowerbound = 0, upperbound = 1)
+end
+
+""
+function variable_voltage(pm::QCWRTriPowerModel, n::Int=pm.cnw; kwargs...)
+    variable_voltage_angle(pm, n; kwargs...)
+    variable_voltage_magnitude(pm, n; kwargs...)
+
+    variable_voltage_magnitude_sqr(pm, n; kwargs...)
+    variable_voltage_product(pm, n; kwargs...)
+
+    variable_voltage_angle_difference(pm, n; kwargs...)
+    variable_voltage_magnitude_product(pm, n; kwargs...)
+    variable_multipliers(pm, n; kwargs...)
+    variable_cosine(pm, n; kwargs...)
+    variable_sine(pm, n; kwargs...)
+    variable_current_magnitude_sqr(pm, n; kwargs...)
+end
+
+""
+function constraint_voltage(pm::QCWRTriPowerModel, n::Int)
+    v = pm.var[:nw][n][:vm]
+    t = pm.var[:nw][n][:va]
+
+    td = pm.var[:nw][n][:td]
+    si = pm.var[:nw][n][:si]
+    cs = pm.var[:nw][n][:cs]
+
+    w = pm.var[:nw][n][:w]
+    wr = pm.var[:nw][n][:wr]
+    lambdar = pm.var[:nw][n][:lambdar]
+    wi = pm.var[:nw][n][:wi]
+    lambdai = pm.var[:nw][n][:lambdai]
+
+    for (i,b) in pm.ref[:nw][n][:bus]
+        relaxation_sqr(pm.model, v[i], w[i])
+    end
+
+    for bp in keys(pm.ref[:nw][n][:buspairs])
+        i,j = bp
+        @constraint(pm.model, t[i] - t[j] == td[bp])
+
+        relaxation_sin(pm.model, td[bp], si[bp])
+        relaxation_cos(pm.model, td[bp], cs[bp])
+        relaxation_trilinear(pm.model, v[i], v[j], cs[bp], wr[bp], lambdar[bp,:])
+        relaxation_trilinear(pm.model, v[i], v[j], si[bp], wi[bp], lambdai[bp,:])
+
+        # this constraint is redudant and useful for debugging
+        #relaxation_complex_product(pm.model, w[i], w[j], wr[bp], wi[bp])
+   end
+
+   for (i,branch) in pm.ref[:nw][n][:branch]
+        pair = (branch["f_bus"], branch["t_bus"])
+        buspair = pm.ref[:nw][n][:buspairs][pair]
+
+        # to prevent this constraint from being posted on multiple parallel branchs
+        if buspair["branch"] == i
+            constraint_power_magnitude_sqr(pm, n, i)
+            constraint_power_magnitude_link(pm, n, i)
+        end
+    end
+
+end
+
+function add_bus_voltage_setpoint(sol, pm::QCWRTriPowerModel)
+    add_setpoint(sol, pm, "bus", "vm", :vm)
```

Comment:
A better fix to this issue would be to generalize the current `add_bus_voltage_setpoint(sol, pm::QCWRPowerModel)` to a function like `add_bus_voltage_setpoint{T <: QCWRForm}(sol, pm::GenericPowerModel{T})`